x-pack/lib/filters/geoip/database_manager.rb

```diff
@@ -30,139 +30,208 @@ module LogStash module Filters module Geoip class DatabaseManager
   include LogStash::Util::Loggable
   include LogStash::Filters::Geoip::Util
 
-  #TODO remove vendor_path
-  def initialize(geoip, database_path, database_type, vendor_path)
-    @geoip = geoip
-    self.class.prepare_cc_db
-    @mode = database_path.nil? ? :online : :offline
-    @mode = :disabled # This is a temporary change that turns off the database manager until it is ready for general availability.
-    @database_type = database_type
-    @database_path = patch_database_path(database_path)
-
-    if @mode == :online
-      logger.info "By not manually configuring a database path with `database =>`, you accepted and agreed MaxMind EULA. "\
-                  "For more details please visit https://www.maxmind.com/en/geolite2/eula"
+  @@instance = nil
+  @@instance_mutex = Mutex.new
 
-      setup
-      clean_up_database
-      execute_download_job
+  def self.instance
+    return @@instance if @@instance
 
-      # check database update periodically. trigger `call` method
-      @scheduler = Rufus::Scheduler.new({:max_work_threads => 1})
-      @scheduler.every('24h', self)
-    elsif @mode == :disabled
-      # The plan is to use CC database in Logstash 7.x and enable EULA database in 8
-    else
-      logger.info "GeoIP database path is configured manually so the plugin will not check for update. "\
-                  "Keep in mind that if you are not using the database shipped with this plugin, "\
-                  "please go to https://www.maxmind.com/en/geolite2/eula and understand the terms and conditions."
+    @@instance_mutex.synchronize do
+      return @@instance if @@instance
+      @@instance = new
     end
+
+    @@instance
   end
 
-  DEFAULT_DATABASE_FILENAME = %w{
-    GeoLite2-City.mmdb
-    GeoLite2-ASN.mmdb
-  }.map(&:freeze).freeze
+  private_class_method :new
 
-  public
+  private
+  def initialize
+    self.class.prepare_cc_db
+    cc_city_database_path = get_db_path(CITY, CC)
+    cc_asn_database_path = get_db_path(ASN, CC)
+
+    prepare_metadata
+    city_database_path = @metadata.database_path(CITY) || cc_city_database_path
+    asn_database_path = @metadata.database_path(ASN) || cc_asn_database_path
+
+    @triggered = false
+    @trigger_lock = Mutex.new
+    @states = { "#{CITY}" => DatabaseState.new(@metadata.is_eula(CITY),
+                                               city_database_path,
+                                               cc_city_database_path),
+                "#{ASN}" => DatabaseState.new(@metadata.is_eula(ASN),
+                                              asn_database_path,
+                                              cc_asn_database_path) }
+
+    @download_manager = DownloadManager.new(@metadata)
+  end
 
+  protected
   # create data dir, path.data, for geoip if it doesn't exist
   # copy CC databases to data dir
   def self.prepare_cc_db
```

Comment:
```suggestion
  def prepare_cc_db
```
I don't see a need to have this a class method, especially if DatabaseManager becomes a singleton and testing is done as mentioned below in https://github.com/elastic/logstash/pull/12862/files#r646476715